okhttp/src/main/java/okhttp3/internal/platform/AndroidSocketAdapter.kt

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (C) 2019 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.internal.platform
+
+import okhttp3.Protocol
+import java.lang.reflect.InvocationTargetException
+import java.lang.reflect.Method
+import java.nio.charset.StandardCharsets
+import javax.net.ssl.SSLSocket
+import javax.net.ssl.SSLSocketFactory
+import javax.net.ssl.X509TrustManager
+
+open class AndroidSocketAdapter internal constructor(private val packageName: String) {
+  internal lateinit var sslSocketClass: Class<in SSLSocket>
+  internal lateinit var paramClass: Class<*>
+  internal val setUseSessionTickets: Method by lazy {
+    sslSocketClass.getDeclaredMethod("setUseSessionTickets", Boolean::class.javaPrimitiveType)
+  }
+  internal val setHostname by lazy {
+    sslSocketClass.getMethod("setHostname", String::class.java)
+  }
+  internal val getAlpnSelectedProtocol by lazy {
+    sslSocketClass.getMethod("getAlpnSelectedProtocol")
+  }
+  internal val setAlpnProtocols by lazy {
+    sslSocketClass.getMethod("setAlpnProtocols", ByteArray::class.java)
+  }
+
+  open fun isSupported(): Boolean = AndroidPlatform.isSupported
+
+  fun trustManager(sslSocketFactory: SSLSocketFactory): X509TrustManager? {
+    val context: Any? =
+        Platform.readFieldOrNull(sslSocketFactory, paramClass, "sslParameters")
+    val x509TrustManager = Platform.readFieldOrNull(
+        context!!, X509TrustManager::class.java, "x509TrustManager")
+    return x509TrustManager ?: Platform.readFieldOrNull(context, X509TrustManager::class.java,
+        "trustManager")
+  }
+
+  fun matchesSocket(sslSocket: SSLSocket): Boolean {
+    val matches = matchesSocketInternal(sslSocket)
+
+    if (matches) {
+      initSocketClasses(sslSocket.javaClass)
```

Comment:
This is the core of the fix, finding the Gms classes isn't obvious (classpath tricks?), but the actual class can provide it.